src/geometry/dual_quaternion.rs

```diff
@@ -0,0 +1,116 @@
+use crate::{Quaternion, SimdRealField};
+
+/// A dual quaternion.
+///
+/// # Indexing
+///
+/// DualQuaternions are stored as \[..real, ..dual\].
+/// Both of the quaternion components are laid out in `w, i, j, k` order.
+///
+/// ```
+/// # use nalgebra::{DualQuaternion, Quaternion};
+///
+/// let real = Quaternion::new(1.0, 2.0, 3.0, 4.0);
+/// let dual = Quaternion::new(5.0, 6.0, 7.0, 8.0);
+///
+/// let dq = DualQuaternion::from_real_and_dual(real, dual);
+/// assert_eq!(dq[0], 1.0);
+/// assert_eq!(dq[4], 5.0);
+/// assert_eq!(dq[6], 7.0);
+/// ```
+///
+/// NOTE:
+///  As of December 2020, dual quaternion support is a work in progress.
+///  If a feature that you need is missing, feel free to open an issue or a PR.
+///  See https://github.com/dimforge/nalgebra/issues/487
+#[repr(C)]
+#[derive(Debug, Default, Eq, PartialEq, Copy, Clone)]
+#[cfg_attr(feature = "serde-serialize", derive(Serialize, Deserialize))]
+pub struct DualQuaternion<N: SimdRealField> {
+    // [real(w, i, j, k), dual(w, i, j, k)]
+    pub(crate) dq: [N; 8],
```

Comment:
Changed to store internally as an array (instead of storing real and dual part separately) so that we don't need an `if` statement when indexing.

Reply:
I think what you did before was better since it made it more explicit that it is composed of two quaternions. And these two quaternions could be public fields for easier access.

Regarding indexing, you can simply:
- Add `#[repr(C)]` to `DualQuaternion`.
- Add an implementation of `AsRef` and `AsMut`. Something like:

 ```rust
 impl<N: SimdRealField> AsRef<[N; 8]> for DualQuaternion<N> {
     #[inline]
     fn as_ref(&self) -> &[N; 8] {
        unsafe { mem::transmute(self) }
    }
}

// Do something similar for AsMut.
```
and use these in the implementation of `Index` and `IndexMut`.

Reply:
Ah I like your suggestion better, thanks will change.

Question though - why are quaternions stored internally at [i,j,k,w]? I would've expected [w,i,j,k] given that `Quaternion::new(w,i,j,k)`.

No big deal, just curious.

Reply:
> Question though - why are quaternions stored internally at [i,j,k,w]? I would've expected [w,i,j,k] given that Quaternion::new(w,i,j,k).

By similarity with the Eigen library, mostly.